src/language/schema/parser.js

```diff
@@ -0,0 +1,274 @@
+/* @flow */
+/**
+ *  Copyright (c) 2015, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+import { Source } from './../source';
+
+import { TokenKind } from './../lexer';
+
+import {
+  ParseOptions,
+  makeParser,
+  peek,
+  skip,
+  loc,
+  any,
+  many,
+  expect,
+  unexpected,
+  expectKeyword,
+  advance,
+} from './../parserCore';
+
+import {
+  parseName,
+  parseNamedType,
+  parseType,
+} from './../parser';
+
+import type {
+  SchemaDocument,
+  TypeDefinition,
+  FieldDefinition,
+  EnumDefinition,
+  EnumValueDefinition,
+  InterfaceDefinition,
+  ArgumentDefinition,
+  ScalarDefinition,
+  InputObjectDefinition,
+} from './ast';
+
+import {
+  SCHEMA_DOCUMENT,
+  ENUM_DEFINITION,
+  ENUM_VALUE_DEFINITION,
+  TYPE_DEFINITION,
+  INTERFACE_DEFINITION,
+  FIELD_DEFINITION,
+  ARGUMENT_DEFINITION,
+  UNION_DEFINITION,
+  SCALAR_DEFINITION,
+  INPUT_OBJECT_DEFINITION,
+} from './kinds';
+
+export function parseSchema(
+  source: Source | string,
+  options?: ParseOptions
+): SchemaDocument {
+  var sourceObj = source instanceof Source ? source : new Source(source);
+  var parser = makeParser(sourceObj, options || {});
+  return parseSchemaDocument(parser);
+}
+
+function parseNextSchemaDefinition(parser): any {
+  switch (parser.token.value) {
+    case 'type':
+      return parseTypeDefinition(parser);
+    case 'enum':
+      return parseEnumDefinition(parser);
+    case 'interface':
+      return parseInterfaceDefinition(parser);
+    case 'union':
+      return parseUnionDefinition(parser);
+    case 'input':
+      return parseInputObjectDefinition(parser);
+    case 'scalar':
+      return parseScalarDefinition(parser);
+    default:
+      throw unexpected(parser);
+  }
+}
+
+function parseScalarDefinition(parser): ScalarDefinition {
+  var start = parser.token.start;
+  expectKeyword(parser, 'scalar');
+  var name = parseName(parser);
+  var location = loc(parser, start);
+  return {
+    kind: SCALAR_DEFINITION,
+    name: name,
+    loc: location,
+  };
+}
+
+function parseSchemaDocument(parser): SchemaDocument {
```

Comment:
just in terms of ease of reading this file, I'd suggest moving this fn between `parseSchema` and `parseNextSchemaDefinition` (which I may also suggest naming just `parseSchemaDefinition`)